topdown/crypto.go

```diff
@@ -160,12 +164,68 @@ func builtinCryptoSha256(a ast.Value) (ast.Value, error) {
 
 func init() {
 	RegisterFunctionalBuiltin1(ast.CryptoX509ParseCertificates.Name, builtinCryptoX509ParseCertificates)
+	RegisterBuiltinFunc(ast.CryptoX509ParseAndVerifyCertificates.Name, builtinCryptoX509ParseAndVerifyCertificates)
 	RegisterFunctionalBuiltin1(ast.CryptoMd5.Name, builtinCryptoMd5)
 	RegisterFunctionalBuiltin1(ast.CryptoSha1.Name, builtinCryptoSha1)
 	RegisterFunctionalBuiltin1(ast.CryptoSha256.Name, builtinCryptoSha256)
 	RegisterFunctionalBuiltin1(ast.CryptoX509ParseCertificateRequest.Name, builtinCryptoX509ParseCertificateRequest)
 }
 
+func getX509CertsFromBytes(bytes []byte) ([]*x509.Certificate, error) {
+	// if the input is not a PEM string, attempt to decode b64
+	if str := string(bytes); !strings.HasPrefix(str, "-----BEGIN CERTIFICATE-----") {
+		var err error
+		bytes, err = base64.StdEncoding.DecodeString(str)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// attempt to decode input as PEM data
+	p, rest := pem.Decode(bytes)
+	if p != nil && p.Type != blockTypeCertificate {
```

Comment:
❓ We're having three `if p != nil` here, would it perhaps be preferable to go with `if p == nil { /* return something */ }` and go ahead with `p` if not? (The check on line 199 is different, of course.)

Reply:
This actually isn't related to this change. This was the exiting body of `parse_certificates` which I just moved out into its own function so it could be reused by `parse_and_verify_certificates`. I didn't look into the code at all since it was here previously and appears to work. That said, now that I've looked I agree this could be improved.